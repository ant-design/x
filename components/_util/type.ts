export type AnyObject = Record<PropertyKey, any>;

export interface BaseProps {
  /**
   * @desc 自定义前缀
   * @descEN Prefix
   */
  prefixCls?: string;

  /**
   * @desc 自定义根类名
   * @descEN Custom class name
   */
  rootClassName?: string;
}