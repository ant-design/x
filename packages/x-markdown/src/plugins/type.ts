import type { KatexOptions } from 'katex';
import type { TokenizerAndRendererExtension } from 'marked';
import type { ReactNode } from 'react';
import type { SyntaxHighlighterProps } from 'react-syntax-highlighter';

export type LatexOption = {
  katexOptions?: KatexOptions;
  replaceAlignStart?: boolean;
};

type HighlightCodeType = 'root' | 'header' | 'headerTitle' | 'code';
export type HighlightCodeProps = {
  lang?: string;
  children: string;
  header?: ReactNode | null;
  prefixCls?: string;
  style?: React.CSSProperties;
  className?: string;
  highlightProps?: Partial<SyntaxHighlighterProps>;
  // Semantic
  classNames?: Partial<Record<HighlightCodeType, string>>;
  styles?: Partial<Record<HighlightCodeType, React.CSSProperties>>;
};

type MermaidType = 'root' | 'header' | 'graph' | 'code';
export type MermaidProps = {
  children: string;
  header?: ReactNode | null;
  prefixCls?: string;
  style?: React.CSSProperties;
  className?: string;
  highlightProps?: Partial<SyntaxHighlighterProps>;
  // Semantic
  classNames?: Partial<Record<MermaidType, string>>;
  styles?: Partial<Record<MermaidType, React.CSSProperties>>;
};

interface BaseComponentConfig {
  style: React.CSSProperties;
  styles: Record<string, React.CSSProperties>;
  className: string;
  classNames: Record<string, string>;
}

type ComponentConfig<
  CompProps extends Record<PropertyKey, any>,
  PickType extends keyof CompProps = keyof BaseComponentConfig,
> = Pick<CompProps, PickType>;

export interface MarkdownComponentsConfig {
  highlightCode?: ComponentConfig<HighlightCodeProps>;
  mermaid?: ComponentConfig<MermaidProps>;
}

export type PluginsType = {
  /**
   * @desc 渲染数学公式Latex语法。
   * @descEN Rendering mathematical formulas using Latex syntax.
   */
  Latex: (options?: LatexOption) => TokenizerAndRendererExtension[];
  /**
   * @desc 渲染代码高亮。
   * @descEN Highlight the rendering code.
   */
  HighlightCode: (props: HighlightCodeProps) => React.ReactNode;
  /**
   * @desc 渲染 Mermaid 图表。
   * @descEN Rendering the Mermaid Chart.
   */
  Mermaid: (props: MermaidProps) => React.ReactNode;
};
