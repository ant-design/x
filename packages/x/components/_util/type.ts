export type AnyObject = Record<PropertyKey, any>;
export type PrefixKeysType = {
  Ctrl: keyof KeyboardEvent;
  Alt: keyof KeyboardEvent;
  Meta: keyof KeyboardEvent;
  Shift: keyof KeyboardEvent;
};
export type ShortcutKeys<CustomKey = number | 'number'> =
  | [keyof PrefixKeysType, keyof PrefixKeysType, CustomKey]
  | [keyof PrefixKeysType, CustomKey];
