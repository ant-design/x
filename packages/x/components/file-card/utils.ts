export function getSize(size: number) {
  let retSize = size;
  const units = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB'];
  let unitIndex = 0;

  while (retSize >= 1024 && unitIndex < units.length - 1) {
    retSize /= 1024;
    unitIndex++;
  }

  return `${retSize.toFixed(0)} ${units[unitIndex]}`;
}

export function matchExt(suffix: string, ext: string[]) {
  return ext.some((e) => suffix.toLowerCase() === `.${e}`);
}
