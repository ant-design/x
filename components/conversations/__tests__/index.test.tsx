import React from 'react';
import Conversations from '../index';
import { render, fireEvent } from '../../../tests/utils';

import type { Conversation } from '../index';
import mountTest from '../../../tests/shared/mountTest';
import rtlTest from '../../../tests/shared/rtlTest';

const items: Conversation[] = [
  {
    key: 'demo1',
    label: 'What is Ant Design X ?',
    timestamp: 794620800,
    icon: <div id="conversation-test-id">icon</div>,
    group: 'pinned',
  },
  {
    key: 'demo2',
    label: (
      <div>
        Getting Started:{' '}
        <a target="_blank" href="https://ant-design.antgroup.com/index-cn" rel="noreferrer">
          Ant Design !
        </a>
      </div>
    ),
    timestamp: 794620900,
  },
  {
    key: 'demo4',
    label: 'In Docker, use üêë Ollama and initialize',
    timestamp: 794621100,
  },
  {
    key: 'demo5',
    label: 'Expired, please go to the recycle bin to check',
    timestamp: 794621200,
    disabled: true,
  },
];

describe('Conversations Component', () => {
  mountTest(() => <Conversations />);

  rtlTest(() => <Conversations items={items} />);

  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('Conversations component work', () => {
    const { container } = render(<Conversations items={items} />);
    const element = container.querySelector<HTMLUListElement>('.ant-conversations');
    expect(element).toBeTruthy();
    expect(element).toMatchSnapshot();
  });

  it('should handle defaultActiveKey', () => {
    const { getByText } = render(<Conversations items={items} defaultActiveKey="demo1" />);
    const activeItem = getByText('What is Ant Design X ?');
    expect(activeItem.parentNode).toHaveClass('ant-conversations-item-active');
  });

  it('should handle activeKey', () => {
    const { getByText } = render(<Conversations items={items} activeKey="demo1" />);
    const activeItem = getByText('What is Ant Design X ?');
    expect(activeItem.parentNode).toHaveClass('ant-conversations-item-active');
  });

  it('should trigger onActiveChange', () => {
    const onActiveChange = jest.fn();
    const { getByText } = render(<Conversations items={items} onActiveChange={onActiveChange} />);
    fireEvent.click(getByText('What is Ant Design X ?'));
    expect(onActiveChange).toHaveBeenCalledWith('demo1');
    fireEvent.click(getByText('In Docker, use üêë Ollama and initialize'));
    expect(onActiveChange).toHaveBeenCalledWith('demo4');
    fireEvent.click(getByText('Expired, please go to the recycle bin to check'));
    expect(onActiveChange).toHaveBeenCalledWith('demo4');
  });

  it('should handle menu function', () => {
    const menu = jest.fn().mockReturnValue({
      items: [
        {
          label: 'ÈáçÂëΩÂêç',
          key: 'mod',
        },
        {
          label: 'Âà†Èô§',
          key: 'delete',
          danger: true,
        },
      ],
    });
    const { getByText, container } = render(
      <Conversations
        items={[
          {
            key: 'demo5',
            label:
              'Looooooooooooooooooooooooooonnnnnnnnnnnnnnnnnnnnnnnnnnnnngggggggggggggggggggggg',
            timestamp: 794621200,
          },
        ]}
        menu={menu}
        defaultActiveKey="demo1"
      />,
    );
    expect(menu).toHaveBeenCalled();
    const menuElement = container.querySelector('.ant-conversations-menu-icon');
    expect(menuElement).toBeInTheDocument();
    fireEvent.click(menuElement as HTMLElement);
    expect(getByText('ÈáçÂëΩÂêç')).toBeInTheDocument();
    expect(getByText('Âà†Èô§')).toBeInTheDocument();
  });

  it('should group items when groupable is true', () => {
    const { getByText } = render(<Conversations items={items} groupable />);
    expect(getByText('pinned')).toBeInTheDocument();
  });

  it('should use custom group title component', () => {
    const { getByText } = render(
      <Conversations items={items} groupable={{ title: (group) => <div>{group}</div> }} />,
    );
    expect(getByText('pinned')).toBeInTheDocument();
  });

  it('should sort groups when groupable.sort is provided', () => {
    const sort = jest.fn().mockReturnValue(0);
    render(<Conversations items={items} groupable={{ sort }} />);
    expect(sort).toHaveBeenCalled();
  });

  it('should not group items when groupable is false', () => {
    const { queryByText } = render(<Conversations items={items} groupable={false} />);
    expect(queryByText('pinned')).not.toBeInTheDocument();
  });
});
