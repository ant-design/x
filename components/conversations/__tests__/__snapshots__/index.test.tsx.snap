// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conversations Component Conversations component work 1`] = `
<ul
  class="ant-conversations"
>
  <li
    class="ant-conversations-item"
  >
    <div
      class="ant-conversations-icon"
    >
      <div
        id="conversation-test-id"
      >
        icon
      </div>
    </div>
    <span
      aria-label="What is Ant Design X ?"
      class="ant-typography ant-typography-ellipsis ant-conversations-label"
    >
      What is Ant Design X ?
    </span>
  </li>
  <li
    class="ant-conversations-item"
  >
    <span
      class="ant-typography ant-typography-ellipsis ant-conversations-label"
    >
      <div>
        Getting Started:
        <a
          href="https://ant-design.antgroup.com/index-cn"
          rel="noreferrer"
          target="_blank"
        >
          Ant Design !
        </a>
      </div>
    </span>
  </li>
  <li
    class="ant-conversations-item"
  >
    <span
      aria-label="In Docker, use üêë Ollama and initialize"
      class="ant-typography ant-typography-ellipsis ant-conversations-label"
    >
      In Docker, use üêë Ollama and initialize
    </span>
  </li>
  <li
    class="ant-conversations-item ant-conversations-item-disabled"
  >
    <span
      aria-label="Expired, please go to the recycle bin to check"
      class="ant-typography ant-typography-ellipsis ant-conversations-label"
    >
      Expired, please go to the recycle bin to check
    </span>
  </li>
</ul>
`;

exports[`Conversations Component rtl render component should be rendered correctly in RTL direction 1`] = `
<ul
  class="ant-conversations ant-conversations-rtl"
>
  <li
    class="ant-conversations-item"
  >
    <div
      class="ant-conversations-icon"
    >
      <div
        id="conversation-test-id"
      >
        icon
      </div>
    </div>
    <span
      aria-label="What is Ant Design X ?"
      class="ant-typography ant-typography-rtl ant-typography-ellipsis ant-conversations-label"
    >
      What is Ant Design X ?
    </span>
  </li>
  <li
    class="ant-conversations-item"
  >
    <span
      class="ant-typography ant-typography-rtl ant-typography-ellipsis ant-conversations-label"
    >
      <div>
        Getting Started:
        <a
          href="https://ant-design.antgroup.com/index-cn"
          rel="noreferrer"
          target="_blank"
        >
          Ant Design !
        </a>
      </div>
    </span>
  </li>
  <li
    class="ant-conversations-item"
  >
    <span
      aria-label="In Docker, use üêë Ollama and initialize"
      class="ant-typography ant-typography-rtl ant-typography-ellipsis ant-conversations-label"
    >
      In Docker, use üêë Ollama and initialize
    </span>
  </li>
  <li
    class="ant-conversations-item ant-conversations-item-disabled"
  >
    <span
      aria-label="Expired, please go to the recycle bin to check"
      class="ant-typography ant-typography-rtl ant-typography-ellipsis ant-conversations-label"
    >
      Expired, please go to the recycle bin to check
    </span>
  </li>
</ul>
`;
